File: python/src/resp_bench/client/impl/glide_client.py
"""valkey-glide driver using the async GLIDE Python client (``import glide``).

The async client is used deliberately (the sync ``glide_sync`` package is not
used). One ``GlideClient`` is created per connection, honoring the
``client == connection`` invariant shared across engines.
"""

from __future__ import annotations

from ...config.driver_config import DriverConfig
from ..benchmark_client import AsyncBenchmarkClient
from ..timed_result import TimedResult


class GlideBenchmarkClient(AsyncBenchmarkClient):
    def __init__(self) -> None:
        self._client = None

    async def connect(self, host: str, port: int, config: DriverConfig) -> None:
        from glide import (
            GlideClient,
            GlideClientConfiguration,
            GlideClusterClient,
            GlideClusterClientConfiguration,
            NodeAddress,
            ServerCredentials,
        )

        addresses = [NodeAddress(host, port)]

        credentials = None
        if config.auth and (config.auth.get("password") or config.auth.get("username")):
            credentials = ServerCredentials(
                password=config.auth.get("password", ""),
                username=config.auth.get("username"),
            )

        use_tls = config.tls_enabled()
        timeout_ms = config.command_timeout_ms

        if config.is_cluster():
            conf = GlideClusterClientConfiguration(
                addresses=addresses,
                use_tls=use_tls,
                credentials=credentials,
                request_timeout=timeout_ms,
            )
            self._client = await GlideClusterClient.create(conf)
        else:
            conf = GlideClientConfiguration(
                addresses=addresses,
                use_tls=use_tls,
                credentials=credentials,
                request_timeout=timeout_ms,
            )
            self._client = await GlideClient.create(conf)

    async def ping(self) -> TimedResult:
        return await self._measure(lambda: self._client.ping())

    async def get(self, key: str) -> TimedResult:
        return await self._measure(lambda: self._client.get(key))

    async def set(self, key: str, value: bytes) -> TimedResult:
        return await self._measure(lambda: self._client.set(key, value))

    async def close(self) -> None:
        if self._client is not None:
            await self._client.close()

    def driver_version(self) -> str:
        try:
            from importlib.metadata import version

            return version("valkey-glide")
        except Exception:  # noqa: BLE001 - version is best-effort metadata
            return "unknown"

File: node/src/client/impl/glideClient.ts
/**
 * valkey-glide driver (`@valkey/valkey-glide`).
 *
 * One GlideClient per connection, honouring the `client == connection` invariant
 * shared across engines.
 *
 * Two glide-specific shapes to note:
 * - `close()` is **synchronous** (returns void, not a Promise), unlike every
 *   other driver here.
 * - `get()` returns a `string` by default. That is deliberate and matches the
 *   Python engine and the ioredis/iovalkey clients here, so all drivers are
 *   charged for the same UTF-8 decode. Do not switch one driver to bytes.
 */

import type {
  GlideClient,
  GlideClientConfiguration,
  GlideClusterClient,
  GlideClusterClientConfiguration,
  ServerCredentials,
} from '@valkey/valkey-glide';

import type { DriverConfig } from '../../config/driverConfig.js';
import { measure, type BenchmarkClient } from '../benchmarkClient.js';
import { packageVersion } from '../driverVersion.js';
import type { TimedResult } from '../timedResult.js';

const PACKAGE = '@valkey/valkey-glide';

export class GlideBenchmarkClient implements BenchmarkClient {
  private client: GlideClient | GlideClusterClient | null = null;

  async connect(host: string, port: number, config: DriverConfig): Promise<void> {
    const glide = await import('@valkey/valkey-glide');

    const addresses = [{ host, port }];
    const credentials: ServerCredentials | undefined = config.hasAuth()
      ? ({
          password: config.auth?.password ?? '',
          ...(config.auth?.username ? { username: config.auth.username } : {}),
        } as ServerCredentials)
      : undefined;

    const shared = {
      addresses,
      useTLS: config.tlsEnabled(),
      ...(credentials ? { credentials } : {}),
      ...(config.commandTimeoutMs ? { requestTimeout: config.commandTimeoutMs } : {}),
    };

    this.client = config.isCluster()
      ? await glide.GlideClusterClient.createClient(shared as GlideClusterClientConfiguration)
      : await glide.GlideClient.createClient(shared as GlideClientConfiguration);
  }

  private requireClient(): GlideClient | GlideClusterClient {
    if (this.client === null) throw new Error('glide client is not connected');
    return this.client;
  }

  async ping(): Promise<TimedResult<string>> {
    const client = this.requireClient();
    return measure(async () => String(await client.ping()));
  }

  async get(key: string): Promise<TimedResult<string>> {
    const client = this.requireClient();
    return measure(async () => {
      const value = await client.get(key);
      return value === null ? null : String(value);
    }) as Promise<TimedResult<string>>;
  }

  async set(key: string, value: Buffer): Promise<TimedResult<string>> {
    const client = this.requireClient();
    return measure(async () => String(await client.set(key, value)));
  }

  async close(): Promise<void> {
    // Synchronous in glide -- there is nothing to await.
    this.client?.close();
    this.client = null;
  }

  driverVersion(): string {
    return packageVersion(PACKAGE);
  }
}
